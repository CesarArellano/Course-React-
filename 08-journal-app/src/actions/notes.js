import Swal from 'sweetalert2';

import { db } from "../firebase/firebase-config";
import { fileUpload } from '../helpers/fileUpload';
import { loadNotes } from "../helpers/loadNotes";
import { types } from "../types/types";

export const startNewNote =  () => {
  return async ( dispatch, getState ) => {
    const { auth:{ uid }} = getState();
    const newNote = {
      title: '',
      body: '',
      date: new Date().getTime()
    }
    
    const docRef = await db.collection(`${ uid }/journal/notes`).add( newNote );
    dispatch( activeNote(docRef.id, newNote) );
  }
}

export const activeNote = ( id, note ) => ({
  type: types.notesActive,
  payload: {
    id,
    ...note
  }
});

export const startLoadingNotes = ( uid ) => {
  return async (dispatch) => {
    const notes = await loadNotes(uid);
    dispatch(setNote(notes));
  };
};

export const setNote = ( notes ) => ({
  type: types.notesLoad,
  payload: notes
});


export const startSaveNote = ( note ) => {
  return async (dispatch, getState) => {
    const { auth:{ uid } } = getState();

    if( !note.url ) {
      delete note.url;
    }

    const noteToFirestore = { ...note };
    delete noteToFirestore.id;
    try {
      await db.doc(`${ uid }/journal/notes/${ note.id }`).update( noteToFirestore );
      dispatch( refreshNote(note.id, noteToFirestore) );
      Swal.fire('Saved', note.title, 'success');
    } catch(e) {
      Swal.fire('Error', e, 'error');
    }
    
  };
}

export const refreshNote = (id, note) => ({
  type: types.notesUpdated,
  payload: {
    id,
    note: {
      id,
      ...note
    }
  }
});

export const startUploading = ( file ) => {
  return async (dispatch, getState) => {
    const { active: activeNote } = getState().notes;
    const fileUrl = await fileUpload(file);
    console.log(fileUrl);
  }
}
