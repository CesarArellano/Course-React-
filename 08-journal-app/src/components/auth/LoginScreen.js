import React from 'react'
import { Link } from 'react-router-dom'

export const LoginScreen = () => {
  return (
    <>
      <div className="center-box">
        <img 
          src="https://image.flaticon.com/icons/png/512/3235/3235042.png"
          alt="Journal App Icon" 
          title="App Icon"
          className="img-icon"
        />
      </div>
      <h3 className="auth__title">Login</h3>
      <form>
        <input
          type="text"
          placeholder="Email"
          name="email"
          autoComplete="off"
          className="auth__input"
        />

        <input 
          type="password"
          placeholder="Password"
          name="password"
          className="auth__input"
        />

        <div className="center-box">
          <button
            type="submit"
            className="btn btn-primary mb-5"
          >
            Login
          </button>
        </div>
        <hr />
        <div className="auth__social-networks">
          <p>Login with social networks</p>
          <div className="google-btn">
              <div className="google-icon-wrapper">
                  <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
              </div>
              <p className="btn-text">
                  <b>Sign in with google</b>
              </p>
          </div>
        </div>
          <Link 
            to="/auth/register"
            className="link"
          >
            Create New Account
          </Link>
      </form>
    </>
  )
}
