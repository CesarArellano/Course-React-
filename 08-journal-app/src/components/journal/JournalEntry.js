import React from 'react'

export const JournalEntry = () => {
  return (
    <div className="journal__entry pointer">
      <div 
        className="journal__entry-picture"
        style={{
          backgroundSize: 'cover',
          background: 'url(https://t.ipadizate.es/2020/06/macOS-Big-Sur-Vector-Wave-Wallpaper-iDownloadBlog.jpg)'
          
        }}
      >
      </div>

      <div className="journal__entry-body">
        <p className="journal__entry-title">A new day</p>
        <p className="journal__entry-content">
        Lorem Ipsum is simply dummy text of the printing and typesetting industry.
        </p>
      </div>

      <div className="journal__entry-date-box">
        <span>Monday</span>
        <h4>28</h4>
      </div>
    </div>
  )
}
