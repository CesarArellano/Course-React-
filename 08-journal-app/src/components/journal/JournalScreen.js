import React from 'react'

export const JournalScreen = () => {
  return (
    <div className="container">
      <h1>Journal Screen</h1>
      <hr />
    </div>
  )
}
