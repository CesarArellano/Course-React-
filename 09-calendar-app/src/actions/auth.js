import { fetchWithoutToken } from "../helpers/fetch";
import { types } from "../types/types";

export const startLogin = ( email, password ) => {
  return async ( dispatch ) => {
    const resp = await fetchWithoutToken('auth', { email, password }, 'POST');
    const decodedData = await resp.json();
    
    if( decodedData.ok ) {
      localStorage.setItem( 'token', decodedData.token );
      localStorage.setItem( 'token-init-date', new Date().getTime() );
      dispatch( login({
        uid: decodedData.uid,
        name: decodedData.name,
      }));
    }
  }
};

const login = ( user ) => ({
  type: types.authLogin,
  payload: user,
});