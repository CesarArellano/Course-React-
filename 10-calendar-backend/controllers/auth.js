const { response } = require('express');
const User = require('../models/User');

const createUser = async ( req, res = response ) => {
  
  const { email, password } = req.body;
  try {
    let user = await User.findOne({ email })

    if( user ) {
      return res.status(400).json({
        ok: false,
        msg: 'El usuario ya existe con ese correo.'
      });
    }

    user = new User( req.body );
    user.save();
    
    res.status(201).json({
      ok: true,
      uid: user.id,
      name: user.name
    });

  } catch(error) {
    console.log(error);
    res.status(500).json({
      ok: false,
      msg: 'Please talk to the admin'
    });
  }
  
}

const userLogin = ( req, res = response ) => {
  const { email, password } = req.body;

  res.json({
    ok: true,
    msg: 'login',
    email,
    password
  });

};

const revalidateToken = ( req, res = response ) => {
  res.json({
    ok: true,
    msg: 'renew'
  });
};

module.exports = {
  createUser,
  userLogin,
  revalidateToken
};