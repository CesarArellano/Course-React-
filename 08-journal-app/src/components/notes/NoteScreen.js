import React, { useEffect, useRef } from 'react'
import { useDispatch, useSelector } from 'react-redux';
import { activeNote, startDeleting } from '../../actions/notes';
import { useForm } from '../../hooks/useForm';
import { NotesAppBar } from './NotesAppBar'

export const NoteScreen = () => {
  
  const dispatch = useDispatch();
  const { active:note } = useSelector( state => state.notes );
  const [ formValues, handleInputChange, reset ] = useForm( note );
  const { body, title, id } = formValues;

  const activeId = useRef( note.id );
  
  useEffect(() => {
    if( note.id !== activeId.current ) {
      reset(note);
      activeId.current = note.id;
    }
  }, [reset, note]);

  useEffect(() => {
    dispatch( activeNote(formValues.id, { ...formValues } ) );
  }, [dispatch, formValues]);

  const handleDeleteNote = () => {
    dispatch( startDeleting(id) );
  }

  return (
    <div className="notes__main-content">
      <NotesAppBar />
      <div className="notes__content">
        <input
          type="text"
          placeholder="Some awesome title"
          className="notes__title-input"
          name="title"
          value={ title }
          onChange={ handleInputChange }
        />
        <textarea
          placeholder="What happened today"
          className="notes__textarea"
          name="body"
          value={ body }
          onChange={ handleInputChange }
        ></textarea>

        {
          (note.url) 
          && (
            <div className="notes__image">
              <img 
                src={ note.url }
                alt="tree"
              />
            </div>
          )
        }
      </div>
      <button
        className="btn btn-danger"
        onClick={ handleDeleteNote }
      >
        Delete
      </button>
    </div>
  )
}
