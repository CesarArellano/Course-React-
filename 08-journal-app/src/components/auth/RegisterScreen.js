import React from 'react'
import { Link } from 'react-router-dom'
import { useDispatch, useSelector } from 'react-redux';
import { useForm } from '../../hooks/useForm'
import validator from 'validator';
import { removeError, setError } from '../../actions/ui';
import { startRegisterWithEmailPasswordName } from '../../actions/auth';

export const RegisterScreen = () => {
  const dispatch = useDispatch();
  const { ui:{ msgError } } = useSelector( state => state); // Recover state data
  
  const [ formValues, handleInputChange ] = useForm({
    name: 'CÃ©sar',
    email: 'cesarmauricio.arellano@gmail.com',
    password: '12345678',
    password2: '12345678'
  });

  const { name, email, password, password2 } = formValues;
  
  const handleRegister = (e) => {
    e.preventDefault();
    
    if( isFormValid() ) {
      dispatch( startRegisterWithEmailPasswordName( email, password, name) )
    }
  };

  const isFormValid = () => {
    if( name.trim().length === 0 ) {
      dispatch(setError('Name is required'));
      return false;
    } else if( !validator.isEmail(email) ){
      dispatch(setError('Email is not valid'));
      return false;
    } else if ( password !== password2 || password.length < 8 ) {
      dispatch(setError('Password should be at least 8 characters and match each other'));
      return false;
    }
    dispatch( removeError() );
    return true;
  };

  return (
    <>
      <div className="center-box">
        <img 
          src="https://image.flaticon.com/icons/png/512/3235/3235042.png"
          alt="Journal App Icon" 
          title="App Icon"
          className="img-icon"
        />
      </div>
      <h3 className="auth__title">Register</h3>
      <form onSubmit={ handleRegister }>
        {
          ( msgError ) && <div className="auth__alert-error animate__animated animate__shakeX">
            { msgError }
          </div>
        }
        
        <input
          type="text"
          placeholder="Name"
          name="name"
          autoComplete="off"
          className="auth__input"
          value={ name }
          onChange={ handleInputChange }
          required
        />
        <input
          type="text"
          placeholder="Email"
          name="email"
          autoComplete="off"
          className="auth__input"
          value={ email }
          onChange={ handleInputChange }
        />

        <input 
          type="password"
          placeholder="Password"
          name="password"
          className="auth__input"
          value={ password }
          onChange={ handleInputChange }
        />

        <input 
          type="password"
          placeholder="Confirm password"
          name="password2"
          className="auth__input"
          value={ password2 }
          onChange={ handleInputChange }
        />

        <div className="center-box">
          <button
            type="submit"
            className="btn btn-primary mb-5"
          >
            Register
          </button>
        </div>
          <Link 
            to="/auth/login"
            className="link"
          >
            Already registered?
          </Link>
      </form>
    </>
  )
}
