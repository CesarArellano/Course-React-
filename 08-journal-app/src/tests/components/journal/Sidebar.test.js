import '@testing-library/jest-dom';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';

import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';

import { Sidebar } from '../../../components/journal/Sidebar';
import { startLogout } from '../../../actions/auth';
import { startNewNote } from '../../../actions/notes';

jest.mock('../../../actions/auth', () => ({
  startLogout: jest.fn()
}));

jest.mock('../../../actions/notes', () => ({
  startNewNote: jest.fn()
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initState = {
  auth: {
    name: 'CÃ©sar'
  },
  notes: {
    notes: []
  }
}

const store = mockStore(initState);
store.dispatch = jest.fn();

describe('Testing in <Sidebar />', () => {
  
  const wrapper = mount(
    <Provider store={ store }>
      <Sidebar />
    </Provider>
  );

  test('should show the component correctly', () => {
    
    expect(wrapper).toMatchSnapshot();
  });

  test('should call the startLogout', () => {
    wrapper.find('.btn').simulate('click');
    expect( startLogout ).toHaveBeenCalled();
  })

  test('should call the startNewNote', () => {
    wrapper.find('.journal__new-entry').simulate('click');
    expect( startNewNote ).toHaveBeenCalled();
  })
  
  
})
