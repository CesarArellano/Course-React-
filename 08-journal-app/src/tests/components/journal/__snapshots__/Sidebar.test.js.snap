// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing in <Sidebar /> should show the component correctly 1`] = `
<aside
  className="journal__sidebar"
>
  <div
    className="journal__sidebar-navbar"
  >
    <h3
      className="mt-5"
    >
      <i
        className="fa fa-moon"
      />
      <span>
        CÃ©sar
      </span>
    </h3>
    <button
      className="btn"
      onClick={[Function]}
    >
      Logout
    </button>
  </div>
  <div
    className="journal__new-entry"
    onClick={[Function]}
  >
    <i
      className="fa fa-calendar-plus fa-5x"
    />
    <p>
       New Entry
    </p>
  </div>
  <div
    className="journal__entries"
  />
</aside>
`;
